File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.VirtualProxyExample/ImageProxy.cs
namespace ProxyDesignPattern.VirtualProxyExample;

/// <summary>
/// Virtual Proxy - Controls access to the real image.
/// Creates the real image only when Display() is called (lazy loading).
/// Thread-safe implementation using double-check locking.
/// </summary>
public class ImageProxy : IImage
{
    private readonly string _fileName;
    private readonly object _lock = new();
    private HighResolutionImage? _realImage;

    public ImageProxy(string fileName)
    {
        _fileName = fileName;
        Console.WriteLine($"[Proxy] Proxy created for: {_fileName}");
    }

    public void Display()
    {
        // Double-check locking for thread-safe lazy loading
        if (_realImage is null)
        {
            lock (_lock)
            {
                if (_realImage is null)
                {
                    Console.WriteLine($"[Proxy] First access - loading real image...");
                    _realImage = new HighResolutionImage(_fileName);
                }
            }
        }
        
        _realImage.Display();
    }

    public bool IsLoaded => _realImage is not null;
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.VirtualProxyExample/src/ImageProxy.ts
import { HighResolutionImage } from "./HighResolutionImage";
import { IImage } from "./IImage";

/**
 * Virtual Proxy - Controls access to the real image.
 * Creates the real image only when display() is called (lazy loading).
 */
export class ImageProxy implements IImage {
  private _realImage: HighResolutionImage | null = null;
  private _loading = false;

  constructor(private readonly _fileName: string) {
    console.log(`[Proxy] Proxy created for: ${_fileName}`);
  }

  async display(): Promise<void> {
    if (this._realImage === null && !this._loading) {
      this._loading = true;
      console.log(`[Proxy] First access - loading real image...`);
      this._realImage = await HighResolutionImage.create(this._fileName);
      this._loading = false;
    }

    await this._realImage!.display();
  }

  get isLoaded(): boolean {
    return this._realImage !== null;
  }
}
